src/ray/raylet/scheduling_resources.cc

```diff
@@ -257,12 +257,15 @@ void ResourceIds::Release(const ResourceIds &resource_ids) {
     if (fractional_pair_it == fractional_ids_.end()) {
       fractional_ids_.push_back(fractional_pair_to_return);
     } else {
+      RAY_CHECK(fractional_pair_it->second < 1);
       fractional_pair_it->second += fractional_pair_to_return.second;
-      RAY_CHECK(fractional_pair_it->second <= 1);
```

Comment:
This should probably be `RAY_CHECK(fractional_pair_it->second < 1 + std::numeric_limits<double>::epsilon());`

I believe @romilbhardwaj is making this change in https://github.com/ray-project/ray/pull/4555.

However, all of the imprecision is just a temporary solution. There are too many places that we'd have to add `epsilon` in order to make things correct. @williamma12 is working on a solution that simply makes the fractional resource bookkeeping exact (instead of approximate) by using integers instead of doubles (where the integer can represent 1/1000th of a resource).

Reply:
The problem in this case isn't precision. The problem is that the remaining value and returned value can add up to be larger than 1. E.g., returning 0.3 to existing 0.8 will be 1.1 in total.

Reply:
Yes, @raulchen is right. From my observation in `test_actor.py::test_resource_assignment`, there will be 1.1 which is generated by adding 0.2 to 0.9.

Reply:
@raulchen @guoyuhong that shouldn't be possible. If that happens, then it is a bug.

Each fractional resource corresponds to a specific resource ID. Suppose there are two GPUs with IDs 0 and 1, and suppose there are two tasks that require 0.2 and 0.9 GPUs.

- The raylet starts with (ID1, 1), (ID2, 1).
- The first task gets scheduled, so the raylet has (ID1, 0.8), (ID2, 1), and the first worker has (ID1, 0.2).
- The second task gets scheduled, so the raylet has (ID1, 0.8), (ID2, 0.1), the first worker has (ID1, 0.2) and the second worker has (ID2, 0.9)
- The first task finishes, so the raylet has (ID1, 1), (ID2, 0.1), and the second worker has (ID2, 0.9)
- The second task finishes, so the raylet has (ID1, 1), (ID2, 1)

We won't add the 0.2 and the 0.9 together because they correspond to different IDs. It isn't possible for one worker to have (ID1, 0.2) and another worker to have (ID1, 0.9) because the total quantity of a given ID is 1.

Does my notation make sense?

Reply:
@robertnishihara Thanks for the information. I looked at the code again. The `resource_id` is `int64_t` not a `string`. I may misunderstand the code. I run the test several time and cannot repro the case that adding 0.2 to 0.9, which is strange.
Sorry to this bad change. Shall I revert the change or this will be fixed in #4555 by @romilbhardwaj?

Reply:
Thanks @guoyuhong, we'll fix it in #4555.